File: src/Services/AppointmentService/Application/Auth/Roles.cs
namespace AppointmentService.Application.Auth;

/// <summary>
/// Имена realm-ролей Keycloak. Значения должны совпадать с ролями в IdP.
/// </summary>
public static class Roles
{
    public const string Admin = "admin";
    public const string Doctor = "doctor";
    public const string Patient = "patient";

    /// <summary>Для [Authorize(Roles = ...)] — несколько ролей через запятую.</summary>
    public const string PatientOrDoctor = Patient + "," + Doctor;
}

File: src/Web.Generated/src/auth/roles.ts
/** Достает realm-роли из access token Keycloak (claim realm_access.roles). */
export function getRealmRolesFromAccessToken(accessToken: string | undefined): string[] {
  if (!accessToken) return []

  try {
    const payloadPart = accessToken.split('.')[1]
    if (!payloadPart) return []

    const json = atob(payloadPart.replace(/-/g, '+').replace(/_/g, '/'))
    const payload = JSON.parse(json) as {
      realm_access?: { roles?: string[] }
    }

    return payload.realm_access?.roles ?? []
  } catch {
    return []
  }
}

export function hasAdminRole(roles: string[]): boolean {
  return roles.includes('admin')
}

export function hasDoctorRole(roles: string[]): boolean {
  return roles.includes('doctor')
}

export function hasPatientRole(roles: string[]): boolean {
  return roles.includes('patient')
}

/** Приоритет: admin → doctor → patient. */
export function homePathForRoles(roles: string[]): string {
  if (hasAdminRole(roles)) return '/admin'
  if (hasDoctorRole(roles)) return '/doctor'
  if (hasPatientRole(roles)) return '/patient'
  return '/access-denied'
}
